myapp/main.py

```diff
@@ -0,0 +1,490 @@
+            lab_id = r['lab_id']
+            row_name = r['row_name']
+            row_id = Generate_row_id(lab_id, row_name)
+            data_id = Generate_data_id(row_id, student_name)
+            db_session.add(schema.Lab_data(row_id=row_id, data_id=data_id,
+                                           student_name=student_name,
+                                           row_data=r['row_data']))
+    db_session.commit()
+    # return success to the ajax call from flask
+    return jsonify(success=True, data=jsonData)
+
+
+# 4.Admin Side
+# admin homepage
+@app.route('/admin_home')
+def admin_home():
+    return render_template('admin_home.html')
+
+
+# create a new lab/manage current labs
+@app.route('/admin_setup_labs_and_data_access', methods=['GET', 'POST'])
+def admin_setup_labs_and_data_access():
+    # Collect Info of all the existing labs
+    current_labs = {'activated': [], 'downloaded': [], 'unactivated': []}
+    current_labs['activated']
+    current_labs['downloaded']
+    current_labs['unactivated']
+
+    db_session = db.get_session()
+    query_activated = db_session.query(schema.Lab_info).filter(
+        schema.Lab_info.lab_status == 'Activated')
+    query_downloaded = db_session.query(schema.Lab_info).filter(
+        schema.Lab_info.lab_status == 'Downloaded')
+    query_unactivated = db_session.query(schema.Lab_info).filter(
+        schema.Lab_info.lab_status == 'Unactivated')
+
+    for lab_a in query_activated:
+        current_labs['activated'].append({'lab_id': lab_a.lab_id,
+                                          'lab_name': lab_a.lab_name,
+                                          'class_name': lab_a.class_name,
+                                          'prof_name': lab_a.prof_name})
+    for lab_d in query_downloaded:
+        current_labs['downloaded'].append({'lab_id': lab_d.lab_id,
+                                           'lab_name': lab_d.lab_name,
+                                           'class_name': lab_d.class_name,
+                                           'prof_name': lab_d.prof_name})
+    for lab_u in query_unactivated:
+        current_labs['unactivated'].append({'lab_id': lab_u.lab_id,
+                                            'lab_name': lab_u.lab_name,
+                                            'class_name': lab_u.class_name,
+                                            'prof_name': lab_u.prof_name})
+
+    if request.method == 'POST':
+        labinfo = {'lab_name':request.form['lab_name'],'class_name':request.form['class_name'],'prof_name':request.form['prof_name'],'lab_desc':request.form['lab_desc'],
+        'lab_questions':request.form['lab_questions']}
+        return redirect(url_for('.admin_setup_labs',labinfo=labinfo))
+    return render_template('admin_setup_labs_and_data_access.html',current_labs=current_labs)
+
+# fill in the row details of a lab
+@app.route('/admin_setup_labs')
+def admin_setup_labs():
+    labinfo = ast.literal_eval(request.args['labinfo'])
```

Comment:
using ast here is surprising-- are you expecting the client to send python literals?